### PR TITLE
make TransposeShape infer shape form both sides

### PR DESCRIPTION
**Pull request by dtracz**
## Description ##
Make the function TransposeShape use both input and output data. Base on this it completes all unknown dimension sizes. If any of them is know on both sides the function checks if they match.

### CODE REVIEW

**Comment by piyushghai** [comment]
@dtracz Can you look into the CI failures ? 
@mxnet-label-bot Add [Operator, Backend, pr-awaiting-review]

File: src/operator/tensor/matrix_op-inl.h [hunk 1]
```diff
@@ -344,19 +344,40 @@ inline bool TransposeShape(const nnvm::NodeAttrs& attrs,
...
+    for (int i = 0; i < out_shp.ndim(); ++i) {
+      ifUnknownAssignElseCheck(ret, i, out_shp[i]);
+      ifUnknownAssignElseCheck(get, shp.ndim()-1-i, out_shp[i]);
+    }
   } else {
-    CHECK_EQ(shp.ndim(), param.axes.ndim());
+    CHECK_EQ(std::max(shp.ndim(), out_shp.ndim()), param.axes.ndim());
     for (int i = 0; i < shp.ndim(); ++i) {
       CHECK(param.axes[i] < static_cast<int64_t>(shp.ndim()));
       ret[i] = shp[param.axes[i]];
```
Comment (ptrendx) [inline src/operator/tensor/matrix_op-inl.h]: Shouldn't you also do something with `get` here? Otherwise it will be all `-1`s in the next loop I think.

Comment (dtracz) [reply src/operator/tensor/matrix_op-inl.h]: To be honest I just missed it, but fact that it was working despite of it realized me, that all 'second assignments' are redundant. The macro SHAPE_ASSIGN_CHECK does it for me at the end.

**Comment by ptrendx** [comment]
@dtracz As discussed offline, please add a small unit test for this fix, then LGTM.

**Pull request merged**